contrib/babelfishpg_tsql/src/pl_exec-2.c

```diff
@@ -236,6 +242,126 @@ exec_tsql_stmt(PLtsql_execstate *estate, PLtsql_stmt *stmt, PLtsql_stmt *save_es
 	return rc;
 }
 
+static int
+exec_stmt_kill(PLtsql_execstate *estate, PLtsql_stmt_kill *stmt)
+{
+	PGPROC *proc;    	
+	Oid	    sysadmin_oid = get_role_oid("sysadmin", false);  // should really use BABELFISH_SYSADMIN in tds_int.h
+	int	    spid = -1;
+	char   *strtol_endptr;
+
+	if (pltsql_explain_only)
+	{
+		StringInfoData query;
+
+		initStringInfo(&query);
+		appendStringInfo(&query, "KILL ");
+		appendStringInfoString(&query, stmt->spid);
+		append_explain_info(NULL, query.data);
+		return PLTSQL_RC_OK;
+	}
+
+	// Do not allow to run KILL inside a transaction
+	if (IsTransactionBlockActive())
+	{
+		ereport(ERROR,
+			(errcode(ERRCODE_ACTIVE_SQL_TRANSACTION),
+				errmsg("%s command cannot be used inside user transactions.", "KILL")));
+	}
+
+	// Require that the user has 'sysadmin' role
+	if (!has_privs_of_role(GetSessionUserId(), sysadmin_oid)) 
+		{	       
+		ereport(ERROR,
+			(errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
+				errmsg("User does not have permission to use the KILL statement")));
+	}
+
+	// Convert SPID string to int
+	// Although the T-SQL grammar only accepts a numeric value, and it must always be present, play it safe
+	Assert(stmt->spid != NULL);     
+	Assert(strlen(stmt->spid) > 0);
+
+	spid = (int) strtol(stmt->spid, &strtol_endptr, 10);
+	if ((spid == 0) && (*strtol_endptr != '\0'))
+	{
+		ereport(ERROR,
+			(errcode(ERRCODE_NUMERIC_VALUE_OUT_OF_RANGE),
+				errmsg("Session ID %s is not valid", stmt->spid)));            
+	}
+	// SPID value must be positive; the T-SQL grammar allows only a non-negative nuber to be specified
+	// Yet, play it safe and test for it
+	// A variable or expression is not allowed and caught in the parser
+	// All other variants of T-SQL KILL are not supported, this is caught in the parser
+	if (spid <= 0)
+	{
+		ereport(ERROR,
+			(errcode(ERRCODE_NUMERIC_VALUE_OUT_OF_RANGE),
+				errmsg("Session ID %d is not valid", spid)));
+	}
+
+	// Verify it is an actually existing process; otherwise we might just be killing any process on the host
+	proc = BackendPidGetProc(spid);
+	if (proc == NULL)
+	{
+		ereport(ERROR,
+			errcode(ERRCODE_NUMERIC_VALUE_OUT_OF_RANGE),
+				errmsg("Process ID %d is not an active process ID", spid));
+	}
+
+	// Do not kill ourselves 
+	if (spid == MyProcPid)
+	{
+		ereport(ERROR,
+			(errcode(ERRCODE_NUMERIC_VALUE_OUT_OF_RANGE),
+				errmsg("Cannot use KILL to kill your own process.")));
+	}		
+
+	// Verify this is a TDS connection, not a PG connection: we should not kill PG connections from T-SQL.
+	// This can be verified by checking the session to be present in sys.dm_exec_sessions or 
+	// sys.dm_exec_connections, which contains T-SQL connections only
+	// (unlike sys.syprocesses which also contains PG connections since this view is also 
+	//  based on pg_locks and pg_stat_activity)
+	{
```

Comment:
Documentation says: System processes and processes running an extended stored procedure can't be ended.
How are we ensuring the part for extended stored procedures?